packages/squiggle-lang/src/rescript/Reducer/Reducer_MathJs/Reducer_MathJs_Parse.res

```diff
@@ -0,0 +1,158 @@
+*/
+module AE = Reducer_Extra_Array
+module JsG = Reducer_Js_Gate
+module Rerr = Reducer_Error
+
+type reducerError = Rerr.reducerError
+
+type rec node = {
+  "type": string,
+  "isNode": bool,
+  "comment": string
+}
+type arrayNode = {
+  ...node,
+  "items": array<node>
+}
+//assignmentNode
+//blockNode
+//conditionalNode
+type constantNode = {
+  ...node,
+  "value": unit
+}
+//functionAssignmentNode
+type functionNode = {
+  ...node,
+  "fn": string,
+  "args": array<node>
+}
+type indexNode = {
+  ...node,
+  "dimensions": array<node>
+}
+type objectNode = {
+  ...node,
+  "properties": Js.Dict.t<node>
+}
+type accessorNode = {
+  ...node,
+  "object": node,
+  "index": indexNode
+}
+type operatorNode = {
+  ...functionNode,
+  "op": string,
+}
+
+//parenthesisNode
+type parenthesisNode = {
+  ...node,
+  "content": node
+}
+//rangeNode
+//relationalNode
+type symbolNode = {
+  ...node,
+  "name": string
+}
+
+external castAccessorNode: node => accessorNode = "%identity"
+external castArrayNode: node => arrayNode = "%identity"
+external castConstantNode: node => constantNode = "%identity"
+external castFunctionNode: node => functionNode = "%identity"
+external castIndexNode: node => indexNode = "%identity"
+external castObjectNode: node => objectNode = "%identity"
+external castOperatorNode: node => operatorNode = "%identity"
+external castOperatorNodeToFunctionNode: operatorNode => functionNode = "%identity"
+external castParenthesisNode: node => parenthesisNode = "%identity"
+external castSymbolNode: node => symbolNode = "%identity"
+
+/*
+  MathJs Parser
+*/
+@module("mathjs") external parse__: string => node = "parse"
+
+let parse = (expr: string): result<node, reducerError> =>
+  try {
+    Ok(parse__(expr))
+  } catch {
+  | Js.Exn.Error(obj) =>
+    RerrJs(Js.Exn.message(obj), Js.Exn.name(obj))->Error
+  }
+
+type mjNode =
+  | MjAccessorNode(accessorNode)
+  | MjArrayNode(arrayNode)
+  | MjConstantNode(constantNode)
+  | MjFunctionNode(functionNode)
+  | MjIndexNode(indexNode)
+  | MjObjectNode(objectNode)
+  | MjOperatorNode(operatorNode)
+  | MjParenthesisNode(parenthesisNode)
+  | MjSymbolNode(symbolNode)
+
+let castNodeType = (node: node) => switch node["type"] {
+  | "AccessorNode" => node -> castAccessorNode -> MjAccessorNode -> Ok
+  | "ArrayNode" => node -> castArrayNode -> MjArrayNode -> Ok
+  | "ConstantNode" => node -> castConstantNode -> MjConstantNode -> Ok
+  | "FunctionNode" => node -> castFunctionNode -> MjFunctionNode -> Ok
+  | "IndexNode" => node -> castIndexNode -> MjIndexNode -> Ok
+  | "ObjectNode" => node -> castObjectNode -> MjObjectNode -> Ok
+  | "OperatorNode" => node -> castOperatorNode -> MjOperatorNode -> Ok
+  | "ParenthesisNode" => node -> castParenthesisNode -> MjParenthesisNode -> Ok
+  | "SymbolNode" => node -> castSymbolNode -> MjSymbolNode -> Ok
+  | _ => Rerr.RerrTodo(`Argg, unhandled MathJsNode: ${node["type"]}`)-> Error
+}
+
+let rec show = (mjNode: mjNode): string => {
```

Comment:
Is this quite right? I don't see where the `rec` is.

Reply:
Yes there is a recursion. Remove it and the code does not run! Because the code is referring to itself inside its definition

Reply:
ok cool, I just didn't read good.